src/docfx/lib/git/GitCommitCacheProvider.cs

```diff
@@ -0,0 +1,93 @@
+    {
+        private const int MaxCommitCacheCountPerFile = 10;
+
+        public static async Task<ConcurrentDictionary<string, Dictionary<(long commit, long blob), (long[] commitHistory, int lruOrder)>>> LoadCommitCache(Repository repo)
+        {
+            Debug.Assert(repo != null);
+            var cacheFilePath = AppData.GetCommitCachePath(repo.Remote);
+            if (!File.Exists(cacheFilePath))
+            {
+                return new ConcurrentDictionary<string, Dictionary<(long commit, long blob), (long[] commitHistory, int lruOrder)>>();
+            }
+            return await ProcessUtility.ReadFile(cacheFilePath, stream =>
+            {
+                var result = new ConcurrentDictionary<string, Dictionary<(long commit, long blob), (long[] commitHistory, int lruOrder)>>();
+                using (var reader = new BinaryReader(stream))
+                {
+                    var fileCount = reader.ReadInt32();
+                    for (var fileIndex = 0; fileIndex < fileCount; fileIndex++)
+                    {
+                        var file = reader.ReadString();
+                        var cacheCount = reader.ReadInt32();
+                        var cachedCommits = result.GetOrAdd(file, _ => new Dictionary<(long, long), (long[], int)>());
+                        for (var cacheIndex = 0; cacheIndex < cacheCount; cacheIndex++)
+                        {
+                            var commit = reader.ReadInt64();
+                            var blob = reader.ReadInt64();
+                            var commitCount = reader.ReadInt32();
+                            var commitHistory = new long[commitCount];
+                            for (var commitIndex = 0; commitIndex < commitCount; commitIndex++)
+                            {
+                                commitHistory[commitIndex] = reader.ReadInt64();
+                            }
+                            cachedCommits.Add((commit, blob), (commitHistory, cacheIndex + 1));
+                        }
+                    }
+                }
+                return result;
+            });
+        }
+
+        public static Task SaveCache(Repository repo, IReadOnlyList<KeyValuePair<string, Dictionary<(long commit, long blob), (long[] commitHistory, int lruOrder)>>> commitCache)
+        {
+            Debug.Assert(repo != null);
+
+            if (commitCache == null)
+            {
+                return Task.CompletedTask;
+            }
+
+            var cacheFilePath = AppData.GetCommitCachePath(repo.Remote);
+            PathUtility.CreateDirectoryFromFilePath(cacheFilePath);
+            return ProcessUtility.WriteFile(cacheFilePath, stream =>
+            {
+                using (var writer = new BinaryWriter(stream))
+                {
+                    writer.Write(commitCache.Count);
+                    foreach (var (file, value) in commitCache)
+                    {
+                        lock (value)
```

Comment:
I am a little worried that this implicit lock is now split across two places: here and [here](https://github.com/dotnet/docfx/pull/3705/files#diff-89b00415785c16d2d2c79b969013484fR161) and [here](https://github.com/dotnet/docfx/pull/3705/files#diff-89b00415785c16d2d2c79b969013484fR109), also, the signature `ConcurrentDictionary<string, Dictionary<(long commit, long blob), (long[] commitHistory, int lruOrder)>>` leaks too much implementation detail, do you want to consolidate and abstract them as well?

Reply:
Maybe we can try to resolve this in a later PR? this PR mainly refactors the scope of GitCommitProvider, split out FileCommitProvider and such on, it is fairly big already.

Reply:
Maybe we can try to resolve this in a later PR? this PR mainly refactors the scope of GitCommitProvider, split out FileCommitProvider and such on, it is fairly big already.